legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabPatchApi.java

```diff
@@ -0,0 +1,357 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.finos.legend.sdlc.server.gitlab.api;
+
+import org.finos.legend.sdlc.domain.model.patch.Patch;
+import org.finos.legend.sdlc.domain.model.version.Version;
+import org.finos.legend.sdlc.domain.model.version.VersionId;
+import org.finos.legend.sdlc.server.domain.api.patch.PatchApi;
+import org.finos.legend.sdlc.server.error.LegendSDLCServerException;
+import org.finos.legend.sdlc.server.gitlab.GitLabConfiguration;
+import org.finos.legend.sdlc.server.gitlab.GitLabProjectId;
+import org.finos.legend.sdlc.server.gitlab.auth.GitLabUserContext;
+import org.finos.legend.sdlc.server.gitlab.tools.GitLabApiTools;
+import org.finos.legend.sdlc.server.gitlab.tools.PagerTools;
+import org.finos.legend.sdlc.server.tools.BackgroundTaskProcessor;
+import org.gitlab4j.api.Constants;
+import org.gitlab4j.api.Pager;
+import org.gitlab4j.api.RepositoryApi;
+import org.gitlab4j.api.models.Branch;
+import org.gitlab4j.api.models.MergeRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.Response;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class GitLabPatchApi extends GitLabApiWithFileAccess implements PatchApi
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(GitLabPatchApi.class);
+
+    @Inject
+    public GitLabPatchApi(GitLabConfiguration gitLabConfiguration, GitLabUserContext userContext, BackgroundTaskProcessor backgroundTaskProcessor)
+    {
+        super(gitLabConfiguration, userContext, backgroundTaskProcessor);
+    }
+
+    private Patch patchBranchToPatch(String projectId, Branch patchBranch)
+    {
+        if (patchBranch != null)
+        {
+           return fromPatchBranchName(projectId, patchBranch.getName());
+        }
+        return null;
+    }
+
+    @Override
+    public Patch newPatch(String projectId, VersionId sourceVersionId)
+    {
+        LegendSDLCServerException.validateNonNull(projectId, "projectId may not be null");
+        LegendSDLCServerException.validateNonNull(sourceVersionId, "source version may not be null");
+
+        GitLabProjectId gitLabProjectId = parseProjectId(projectId);
+
+        Boolean sourceTagExists = false;
+        try
+        {
+            sourceTagExists = GitLabApiTools.tagExists(getGitLabApi(), gitLabProjectId, buildVersionTagName(sourceVersionId));
+        }
+        catch (Exception e)
+        {
+            LOGGER.warn("Error in fetching release tag for version ", sourceVersionId.toVersionIdString(), projectId, e);
+        }
+        if (!sourceTagExists)
+        {
+            throw new LegendSDLCServerException("Release tag for source version " + sourceVersionId.toVersionIdString() + " doesn't exist", Response.Status.BAD_REQUEST);
```

Comment:
We generally should not talk about tags in exception messages.
```suggestion
            throw new LegendSDLCServerException("Source version " + sourceVersionId.toVersionIdString() + " does not exist", Response.Status.BAD_REQUEST);
```